src/matrix_cookbook/3_inverses.lean

```diff
@@ -64,26 +73,121 @@ lemma eq_155 (A B : matrix m m ℂ) : (A * B)⁻¹ = B⁻¹ * A⁻¹ := mul_inv_
 
 /-! ### The Woodbury identity -/
 
-lemma eq_156 : sorry := sorry
-lemma eq_157 : sorry := sorry
-lemma eq_158 : sorry := sorry
+lemma eq_156 (A : matrix m m ℂ) (B : matrix n n ℂ) (C : matrix m n ℂ)
+  {hA: is_unit A.det} {hB: is_unit B.det} {hQ: is_unit (B⁻¹ + Cᵀ ⬝ A⁻¹ ⬝ C).det}: 
+  (A + C⬝B⬝Cᵀ)⁻¹ = A⁻¹ - A⁻¹⬝C⬝(B⁻¹+Cᵀ⬝A⁻¹⬝C)⁻¹⬝Cᵀ ⬝ A⁻¹ :=  
+begin
+  apply sherman_morrison,
+  assumption',
+end
+
+lemma eq_156_hermitianized (A : matrix m m ℂ) (B : matrix n n ℂ) (C : matrix m n ℂ)
+  {hA: is_unit A.det} {hB: is_unit B.det} {hQ: is_unit (B⁻¹ + Cᴴ ⬝ A⁻¹ ⬝ C).det}: 
+  (A + C⬝B⬝Cᴴ)⁻¹ = A⁻¹ - A⁻¹⬝C⬝(B⁻¹+Cᴴ⬝A⁻¹⬝C)⁻¹⬝Cᴴ ⬝ A⁻¹ :=  
+begin
+  apply sherman_morrison,
+  assumption',
+end
+
+lemma eq_157 (A : matrix m m ℂ) (B : matrix n n ℂ) (U : matrix m n ℂ) (V : matrix n m ℂ) 
+  {hA: is_unit A.det} {hB: is_unit B.det} {hQ: is_unit (B⁻¹ + V ⬝ A⁻¹ ⬝ U).det}:
+  (A + U⬝B⬝V)⁻¹ = A⁻¹ - A⁻¹⬝U⬝(B⁻¹+V⬝A⁻¹⬝U)⁻¹⬝V ⬝ A⁻¹ := 
+begin
+  apply sherman_morrison,
+  assumption',
+end
+
+lemma eq_158_hermitianized (P : matrix m m ℂ) (R : matrix n n ℂ) (B : matrix n m ℂ)
+  [invertible P] [invertible R]
+  {hP: matrix.pos_def P} {hR: matrix.pos_def R} : 
+  (P⁻¹ + Bᴴ⬝R⁻¹⬝B)⁻¹⬝Bᴴ⬝R⁻¹ = P⬝Bᴴ⬝(B⬝P⬝Bᴴ + R)⁻¹ := 
+begin
+  -- This is equation 80:
+  -- http://www.stat.columbia.edu/~liam/teaching/neurostat-spr12/papers/hmm/KF-welling-notes.pdf
+
+  have hP_unit := is_unit_iff_ne_zero.2 (pos_def.det_ne_zero hP),
+  have hR_unit := is_unit_iff_ne_zero.2 (pos_def.det_ne_zero hR),
+  have hP_inv_unit := is_unit_nonsing_inv_det P hP_unit,
+  have hR_inv_unit := is_unit_nonsing_inv_det R hR_unit,
+  have hComb_unit: is_unit (R + B ⬝ P ⬝ Bᴴ).det, 
+  { apply is_unit_iff_ne_zero.2, 
+    apply pos_def.det_ne_zero,
+    apply pos_def.add_semidef hR,
+    exact pos_def.hermitian_conj_is_semidef hP, },
+  have : is_unit (R⁻¹⁻¹ + Bᴴᴴ ⬝ P⁻¹⁻¹ ⬝ Bᴴ).det, 
+  { simp only [inv_inv_of_invertible, conj_transpose_conj_transpose],
+    apply hComb_unit, },
+
+  rw add_comm _ R,
+  nth_rewrite 1 ← conj_transpose_conj_transpose B,
+  rw eq_156_hermitianized P⁻¹ R⁻¹ Bᴴ,
+  simp only [inv_inv_of_invertible, conj_transpose_conj_transpose],
+  rw [matrix.sub_mul,  matrix.sub_mul, sub_eq_iff_eq_add], 
+  apply_fun (matrix.mul P⁻¹), 
+  rw matrix.mul_add,
+  repeat {rw ←  matrix.mul_assoc P⁻¹ _ _}, 
+  apply_fun (λ x, x⬝R),  
+  dsimp,
+  rw matrix.add_mul,
+  simp only [inv_mul_of_invertible, matrix.one_mul, 
+    inv_mul_cancel_right_of_invertible],
+  repeat {rw matrix.mul_assoc (Bᴴ⬝(R + B ⬝ P ⬝ Bᴴ)⁻¹)},
+  rw ← matrix.mul_add (Bᴴ⬝(R + B ⬝ P ⬝ Bᴴ)⁻¹), 
+  rw nonsing_inv_mul_cancel_right,
+
+  assumption',
+
+  apply right_mul_inj_of_invertible,
+  apply left_mul_inj_of_invertible,
+end
```

Comment:
I ported this one as #16, using the mathlib-ed versions of many of your theorems; thanks for the proof outline, even though it took over a year to get merged!